src/main/java/uk/gov/justice/digital/delius/service/ReferralService.java

```diff
@@ -111,9 +114,35 @@ public Optional<Nsi> getExistingMatchingNsi(@NotNull final String crn,
         // determine if there is an existing suitable NSI
         var offenderId = offenderService.offenderIdOfCrn(crn).orElseThrow(() -> new BadRequestException("Offender CRN not found"));
 
+        var existingNsis = findReferralNSIByURN(referralId, offenderId);
+        if (existingNsis.isEmpty()) {
+            existingNsis = findReferralNSIByFuzzyMatching(crn, contextName, sentenceId, contractType, startedAt, referralId, offenderId);
```

Comment:
Well, it's not _fuzzy_ matching; more like _categorical data matching_. Should I rename it?